File: app/src/main/java/com/selfawarelab/workouttracker/database/WorkoutDaySerializer.java
package com.selfawarelab.workouttracker.database;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.SerializerProvider;
import com.fasterxml.jackson.databind.ser.std.StdSerializer;
import com.selfawarelab.workouttracker.WorkoutDay;

import java.io.IOException;

public class WorkoutDaySerializer extends StdSerializer<WorkoutDay> {
    public WorkoutDaySerializer() {
        this(null);
    }

    private WorkoutDaySerializer(Class<WorkoutDay> t) {
        super(t);
    }

    @Override
    public void serialize(WorkoutDay value, JsonGenerator jgen, SerializerProvider provider)
            throws IOException, JsonProcessingException {

        jgen.writeStartObject();

        jgen.writeObjectField("workout", value.getWorkout());
        jgen.writeObjectField("dayString", value.getDay());
        jgen.writeNumberField("mDrawable", value.getWorkout().getIcon());
        jgen.writeBooleanField("mIsDisabled", !value.isEnabled());

        jgen.writeEndObject();
    }
}

File: app/src/main/java/com/selfawarelab/workouttracker/database/WorkoutDaySerializer.kt
package com.selfawarelab.workouttracker.database

import com.applandeo.materialcalendarview.EventDay
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.SerializerProvider
import com.fasterxml.jackson.databind.ser.std.StdSerializer
import com.selfawarelab.workouttracker.WorkoutDay
import java.io.IOException
import java.lang.reflect.AccessibleObject.setAccessible

// Don't even bother to write the EventDay. We are going to reinstantiate it everytime because it's too restrictive with property access
class WorkoutDaySerializer private constructor(t: Class<WorkoutDay>?) : StdSerializer<WorkoutDay>(t) {
    constructor() : this(null)

    @Throws(IOException::class, JsonProcessingException::class)
    override fun serialize(value: WorkoutDay, jsonGenerator: JsonGenerator, provider: SerializerProvider) {

        jsonGenerator.let {
            it.writeStartObject()
            it.writeObjectField("workout", value.workout)
            it.writeObjectField("day", value.day)
            it.writeNumberField("icon", value.icon)
            it.writeBooleanField("isEnabled", value.isEnabled)
            it.writeEndObject()
        }
    }
}